components/org.wso2.carbon.identity.conditional.auth.functions.choreo/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/choreo/CallChoreoFunctionImpl.java

```diff
@@ -273,4 +226,321 @@ private String getParentDomainFromUrl(String url) throws URISyntaxException {
+                    Map<String, String> responseBody = this.gson.fromJson(
+                            EntityUtils.toString(response.getEntity()), responseBodyType
+                    );
+                    if (ERROR_CODE_ACCESS_TOKEN_INACTIVE.equals(responseBody.get(CODE))) {
+                        handleExpiredToken();
+                    } else {
+                        LOG.info("Received 401 response from Choreo. Session data key: " + authenticationContext.getContextIdentifier());
+                        this.asyncReturn.accept(authenticationContext, Collections.emptyMap(), Constants.OUTCOME_FAIL);
+                    }
+                } else {
+                    LOG.info("Received non 200 response code from Choreo. Status Code: " + statusCode +
+                            " Session data Key: " + authenticationContext.getContextIdentifier()
+                    );
+                    this.asyncReturn.accept(authenticationContext, Collections.emptyMap(), Constants.OUTCOME_FAIL);
+                }
+            } catch (IOException e) {
+                LOG.error("Error while reading response from Choreo call for session data key: " +
+                        this.authenticationContext.getContextIdentifier(), e
+                );
+                this.asyncReturn.accept(authenticationContext, Collections.emptyMap(), Constants.OUTCOME_FAIL);
+            } catch (Exception e) {
+                LOG.error("Error while processing response from Choreo call for session data key: " +
+                        this.authenticationContext.getContextIdentifier(), e
+                );
+                this.asyncReturn.accept(authenticationContext, Collections.emptyMap(), Constants.OUTCOME_FAIL);
+            }
+        }
+
+        /**
+         * Handles the scenario where the response from the Choreo API call is 401 Unauthorized due to an expired
+         * token. The program will retry the token request flow until it exceeds the specified max request attempt
+         * count.
+         *
+         * @throws SecretManagementException {@link SecretManagementException}
+         * @throws IOException {@link IOException}
+         * @throws FrameworkException {@link FrameworkException}
+         */
+        private void handleExpiredToken() throws SecretManagementException, IOException, FrameworkException {
+
+            if (tokenRequestAttemptCount.get() < MAX_TOKEN_REQUEST_ATTEMPTS) {
+                requestAccessToken(this.authenticationContext.getTenantDomain(), this.connectionMetaData, this);
+                tokenRequestAttemptCount.incrementAndGet();
+            } else {
+                LOG.info("Maximum token request attempt count exceeded for session data key: " + this.authenticationContext.getContextIdentifier());
+                tokenRequestAttemptCount.set(0);
+                this.asyncReturn.accept(authenticationContext, Collections.emptyMap(), OUTCOME_FAIL);
```

Comment:
Check whether there is a proper error message in the FE, when returning with FAIL.

Reply:
When returning with an OUTCOME_FAIL, `onFail` event handler provided to the callChoreo function as an argument gets called. So the developer can define what to do if there was an error in the authentication script. However, if the developer hasn't specified an `onFail` event handler, the authentication flow will continue.